### ISSUE TITLE
Clarify the sequence for loading code/symbols

### ISSUE THREAD

**Issue by simonhaines** [microsoft/debug-adapter-protocol#83]
The sequence diagram in the documentation under the section 'Configuring breakpoint and exception behavior' is confusing for the scenario where code or symbols needs to be loaded into the debug adapter.

The setBreakpoints request resolves to the GDB 'break' command, then the launch request resolves to the 'file' command. These are out-of-order, issuing a 'break' before 'file' results in: No symbol table is loaded.  Use the "file" command.

I'm developing a debug adapter for an interpreter running on a microcontroller and would like some clarity around using the launch request to load code into the interpreter. By the time the launch request is issued, a setBreakpoints request has already been completed. Is it expected the debug adapter cache breakpoints until the launch request, then load the symbols/code, then set the breakpoints? Or is it expected that loading symbols/code is outside the responsibility of the debug adapter and should be handled by an extension contributed command? Or should the code be loaded in the 'initialize' request (which assumes the session is always launching, never attaching)?

**Comment by int19h** [comment]
You might find the diagram in this comment to be more helpful. 
https://github.com/microsoft/vscode/issues/4902#issuecomment-368583522

(perhaps this should be somewhere in the docs?)

**Comment by simonhaines** [comment]
Yes this makes a lot more sense now I realise the initialized event is decoupled from the initialize request/response. I think a brief discussion of 'early' and 'late' initialization (as defined by the linked issue) along with the companion diagram, might be useful. But because this issue is not raised often, it must be just me.

**Comment by int19h** [comment]
I don't think new debuggers are written that often. But for what it's worth, we got it wrong in ours (Python) first time we implemented it, and it was that comment that clarified it. So if you also found it helpful, I think the docs could use clarification in line with the diagram.

**Comment by puremourning** [comment]
Plus one (or as many plusses as I can have) for adding it to the docs. I had to reverse engineer it from VSCode for vimspector and it was not helped that some servers misinterpreted it too.

**Comment by weinand** [comment]
Reopened to improve the documentation.

**Comment by simonhaines** [comment]
Here is a reworking of the init-launch diagram I made to describe the 'late initialization' case. I think the activation box of the Debug Adapter should extend from 'launch request' to 'launch response'.
![init-launch2](https://user-images.githubusercontent.com/18539/70664222-159a0200-1cbe-11ea-8fbc-b3db421427f9.png)

**Comment by puremourning** [comment]
It's actually more complicated than that because 'configurationDone' is optional in the server, so you have to sometimes _always_ send exception breakpoints (and not configurationDone). Or something - I forget the exact details, but it's [a little hairy](https://github.com/puremourning/vimspector/blob/master/python3/vimspector/debug_session.py#L624)

**Comment by simonhaines** [comment]
In any case, the existing diagram (https://microsoft.github.io/debug-adapter-protocol/img/init-launch.png) needs to be changed. The sequence of commands to GDB is out-of-order and confused me at first. GDB does not accept a 'break' command before a 'file' command.

**Comment by weinand** [comment]
@puremourning "configurationDone" is optional because we never want to break any DA with new DAP features. For DAP implementations this doesn't make things complicated or "hairy": All new DAP implementations should only use "configurationDone" and set the capability "supportsConfigurationDoneRequest" to true. But for clients you are right: they have to support both the old and the new style. 

@simonhaines The gdb example shown in the diagram is hypothetical. I never checked whether 'break' is actually allowed before the 'file' command.

But the important information is that the debug adapter and not the client (e.g. VS Code) is in control when it comes to initialising the backend. So the debug adapter starts the initialization sequence when it is ready to accept setBreakpoint etc. requests by sending an initialized event.
That eliminates the need to buffer this information.
